File: Editorial/3583_Count_Special_Triplets/CSharp/2_Enumeration_Binary_Search.cs
using System;
using System.Collections.Generic;

public class Solution {
    private const int MOD = 1000000007;

    public int SpecialTriplets(int[] nums) {
        Dictionary<int, List<int>> pos = new Dictionary<int, List<int>>();

        for (int i = 0; i < nums.Length; i++) {
            int v = nums[i];
            if (!pos.ContainsKey(v)) {
                pos[v] = new List<int>();
            }
            pos[v].Add(i);
        }

        int ans = 0;
        for (int i = 1; i < nums.Length - 1; i++) {
            int target = nums[i] * 2;
            if (pos.ContainsKey(target) && pos[target].Count > 1 &&
                pos[target][0] < i) {
                var (l, r) = UpperBound(pos[target], i);
                int leftCount = l;
                int rightCount = r;
                if (nums[i] == 0) {
                    leftCount--;
                }
                ans = (int)((ans + (long)leftCount * rightCount) % MOD);
            }
        }
        return ans;
    }

    private (int, int) UpperBound(List<int> arr, int i) {
        int l = 0, r = arr.Count - 1;
        while (l < r) {
            int mid = l + ((r - l + 1) >> 1);
            if (i >= arr[mid]) {
                l = mid;
            } else {
                r = mid - 1;
            }
        }
        return (l + 1, arr.Count - 1 - l);
    }
}

File: Editorial/3583_Count_Special_Triplets/TypeScript/2_Enumeration_Binary_Search.ts
const MOD = 1e9 + 7;

function specialTriplets(nums: number[]): number {
    const pos = new Map<number, number[]>();

    for (let i = 0; i < nums.length; i++) {
        const v = nums[i];
        if (pos.has(v)) {
            pos.get(v)!.push(i);
        } else {
            pos.set(v, [i]);
        }
    }

    const upperBound = (arr: number[], i: number) => {
        let l = 0,
            r = arr.length - 1;

        while (l < r) {
            const mid = l + ((r - l + 1) >> 1);
            if (i >= arr[mid]) {
                l = mid;
            } else {
                r = mid - 1;
            }
        }
        return [l + 1, arr.length - 1 - l];
    };

    let ans = 0;
    for (let i = 1; i < nums.length - 1; i++) {
        const targetPos = pos.get(nums[i] * 2);
        if (targetPos && targetPos.length > 1 && targetPos[0] < i) {
            let [l, r] = upperBound(targetPos, i);
            if (nums[i] === 0) {
                l--;
            }
            ans += (l * r) % MOD;
            ans %= MOD;
        }
    }
    return ans;
}
